### PR TITLE
PA-1714 : Add library support to enable the SSE-S3 option during the creation of backup location in the torpedo

### PR DESCRIPTION
**Pull request by sn-px**
**Which issue(s) this PR fixes** (optional)
Closes #https://portworx.atlassian.net/browse/PA-1714

**Special notes for your reviewer**:
Add library support to enable the SSE-S3 option during the creation of backup location in the torpedo
Changes to common code which creates backup location to accept SSE TYPE

Logs : https://jenkins.pwx.dev.purestorage.com/job/Users/job/santoshn/job/byoc/58/console

Latest log : https://jenkins.pwx.dev.purestorage.com/job/Users/job/santoshn/job/byoc/60/consoleFull

Have tests both from admin and user for creation of backup location

Screenshot 2023-10-16 at 5 25 55 PM

### CODE REVIEW

File: tests/common.go [hunk 1]
```diff
@@ -9500,3 +9504,21 @@ func GenerateS3BucketPolicy(sid string, encryptionPolicy string, bucketName stri
 
 	return policy, nil
 }
+
+// GetSseS3EncryptionType fetches SSE type for S3 bucket from the environment variable
+func GetSseS3EncryptionType() (api.S3Config_Sse, error) {
+	var sseType api.S3Config_Sse
+	s3SseTypeEnv := os.Getenv("S3_SSE_TYPE")
+	if s3SseTypeEnv != "" {
+		sseDetails, _ := s3utils.GetS3SSEDetailsFromEnv()
```
[the remark this answers is not in the document]
Comment (sn-px) [reply tests/common.go]: done

File: tests/common.go [hunk 2]
```diff
@@ -9500,3 +9510,24 @@ func GenerateS3BucketPolicy(sid string, encryptionPolicy string, bucketName stri
...
+	if s3SseTypeEnv != "" {
+		sseDetails, err := s3utils.GetS3SSEDetailsFromEnv()
+		if err != nil {
+			return sseType, err
+		}
+		switch strings.ToUpper(string(sseDetails.SseType)) {
+		case string(s3utils.SseS3):
+			sseType = api.S3Config_SSE_S3
+		default:
+			return sseType, fmt.Errorf("failed to sse s3 encryption type not valid: [%v]", sseDetails.SseType)
```
Comment (kshithijiyer-px) [inline tests/common.go:9527]: Do we need to pass `api.S3Config_Invalid` here?

File: tests/common.go [hunk 3]
```diff
@@ -9500,3 +9510,24 @@ func GenerateS3BucketPolicy(sid string, encryptionPolicy string, bucketName stri
...
+		}
+	} else {
+		sseType = api.S3Config_Invalid
```
Comment (kshithijiyer-px) [inline tests/common.go:9530]: If it's invalid don't we need some error here?

File: tests/backup/backup_basic_test.go [hunk 4]
```diff
@@ -157,8 +157,8 @@ var _ = BeforeSuite(func() {
 			globalAWSBucketName = fmt.Sprintf("%s-%s", globalAWSBucketPrefix, bucketNameSuffix)
 			CreateBucket(provider, globalAWSBucketName)
 			log.Infof("Bucket created with name - %s", globalAWSBucketName)
-			s3SseTypeEnv := os.Getenv("S3_SSE_TYPE")
-			if s3SseTypeEnv != "" {
+			s3EncryptionPolicy := os.Getenv("S3_ENCRYPTION_POLICY")
```
Comment (kshithijiyer-px) [inline tests/backup/backup_basic_test.go:160]: Can you make changes to .ssh file as well as currently it won't be there? Is this planned in the next PR?

File: tests/common.go [hunk 5]
```diff
@@ -4239,6 +4239,10 @@ func CreateS3BackupLocation(name string, uid, cloudCred string, cloudCredUID str
 	time.Sleep(60 * time.Second)
 	backupDriver := Inst().Backup
 	_, _, endpoint, region, disableSSLBool := s3utils.GetAWSDetailsFromEnv()
+	sseS3EncryptionType, err := GetSseS3EncryptionType()
```
Comment (kshithijiyer-px) [inline tests/common.go:4242]: Add a comment here

File: tests/common.go [hunk 6]
```diff
@@ -4279,6 +4284,10 @@ func CreateS3BackupLocation(name string, uid, cloudCred string, cloudCredUID str
 func CreateS3BackupLocationWithContext(name string, uid, cloudCred string, cloudCredUID string, bucketName string, orgID string, encryptionKey string, ctx context1.Context) error {
 	backupDriver := Inst().Backup
 	_, _, endpoint, region, disableSSLBool := s3utils.GetAWSDetailsFromEnv()
+	sseS3EncryptionType, err := GetSseS3EncryptionType()
```
Comment (kshithijiyer-px) [inline tests/common.go:4287]: Add a comment here

**Review by kshithijiyer-px — CHANGES_REQUESTED** [review]
Minor comments please take a look

File: tests/common.go (hunk 3 above)
Comment (sn-px) [reply tests/common.go:9530]: Its the naming in the px-backup api

File: tests/backup/backup_basic_test.go (hunk 4 above)
Comment (sn-px) [reply tests/backup/backup_basic_test.go:160]: It is already merged as part of the previous PR

File: tests/common.go (hunk 6 above)
Comment (sn-px) [reply tests/common.go:4289]: done

File: tests/common.go (hunk 5 above)
Comment (sn-px) [reply tests/common.go:4243]: done

File: tests/common.go (hunk 2 above)
Comment (sn-px) [reply tests/common.go:9529]: No we need error out , since the encryption type is invalid

File: tests/common.go (hunk 3 above)
Comment (kshithijiyer-px) [reply tests/common.go:9532]: Let's file a bug for this

File: tests/common.go [hunk 7]
```diff
@@ -4279,6 +4285,11 @@ func CreateS3BackupLocation(name string, uid, cloudCred string, cloudCredUID str
 func CreateS3BackupLocationWithContext(name string, uid, cloudCred string, cloudCredUID string, bucketName string, orgID string, encryptionKey string, ctx context1.Context) error {
 	backupDriver := Inst().Backup
 	_, _, endpoint, region, disableSSLBool := s3utils.GetAWSDetailsFromEnv()
+	//Get SSE S3 Encryption Type
```
Comment (kshithijiyer-px) [inline tests/common.go:4288]: Add a space like `// Get`

File: tests/common.go [hunk 8]
```diff
@@ -4239,6 +4239,11 @@ func CreateS3BackupLocation(name string, uid, cloudCred string, cloudCredUID str
 	time.Sleep(60 * time.Second)
 	backupDriver := Inst().Backup
 	_, _, endpoint, region, disableSSLBool := s3utils.GetAWSDetailsFromEnv()
+	//Get SSE S3 Encryption Type
```
Comment (kshithijiyer-px) [inline tests/common.go:4242]: Add a space like `// Get`

**Review by kshithijiyer-px — APPROVED** [review]
Minor nit picks rest looks good to me

File: tests/common.go (hunk 7 above)
Comment (sn-px) [reply tests/common.go]: done

File: tests/common.go (hunk 8 above)
Comment (sn-px) [reply tests/common.go]: done

File: pkg/s3utils/s3utils.go [hunk 9]
```diff
@@ -32,9 +32,9 @@ type Object struct {
 type sseType string
 
 const (
-	sseS3  sseType = "SSE-S3"
-	sseKms sseType = "SSE-KMS"
-	sseC   sseType = "SSE-C"
+	SseS3  sseType = "SSE-S3"
+	SseKms sseType = "SSE-KMS"
+	SseC   sseType = "SSE-C"
```
Comment (mkoppal-px) [inline pkg/s3utils/s3utils.go:37]: What does `C` mean here?

File: tests/common.go [hunk 10]
```diff
@@ -9500,3 +9512,24 @@ func GenerateS3BucketPolicy(sid string, encryptionPolicy string, bucketName stri
...
+// GetSseS3EncryptionType fetches SSE type for S3 bucket from the environment variable
+func GetSseS3EncryptionType() (api.S3Config_Sse, error) {
+	var sseType api.S3Config_Sse
+	s3SseTypeEnv := os.Getenv("S3_SSE_TYPE")
+	if s3SseTypeEnv != "" {
+		sseDetails, err := s3utils.GetS3SSEDetailsFromEnv()
+		if err != nil {
+			return sseType, err
+		}
+		switch strings.ToUpper(string(sseDetails.SseType)) {
```
Comment (mkoppal-px) [inline tests/common.go:9525]: There is no need to convert it to upper case or string

File: pkg/s3utils/s3utils.go (hunk 9 above)
Comment (sn-px) [reply pkg/s3utils/s3utils.go:37]: Custom key

File: tests/common.go (hunk 10 above)
Comment (sn-px) [reply tests/common.go]: done

**Review by mkoppal-px — CHANGES_REQUESTED** [review]
Please analyse the changed required to create S3 backup location on GCP pipeline. If that needs changes then the functions `CreateGCPBackupLocation` and `CreateGCPBackupLocationWithContext` should also have the SSE type.

File: pkg/s3utils/s3utils.go (hunk 9 above)
Comment (mkoppal-px) [reply pkg/s3utils/s3utils.go:37]: Then let us call it `SseCustomKey`. What do you think?

Comment (sn-px) [reply pkg/s3utils/s3utils.go]: modified .

**Review by mkoppal-px — APPROVED** [review]
LGTM

**Pull request merged**